gluten-core/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala

```diff
@@ -181,7 +185,127 @@ case class OffloadJoin() extends OffloadSingleNode with LogLevelUtil {
+      case b: BatchScanExecTransformer =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case p @ ProjectExec(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case p @ ProjectExecTransformer(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case _ => plan.withNewChildren(plan.children.map(addMetadataCol(_, replacedExprs)))
+    }
+  }
+
+  private def genProjectExec(projectExec: ProjectExec): SparkPlan = {
+    def findScanNodes(plan: SparkPlan): Seq[SparkPlan] = {
+      plan match {
+        case f @ (_: FileSourceScanExec | _: AbstractFileSourceScanExec |
+            _: DataSourceV2ScanExecBase) =>
+          Seq(f)
+        case o => o.children.flatMap(findScanNodes)
+      }
+    }
+
+    if (!TransformHints.isNotTransformable(projectExec)) {
+      logWarning(s"Columnar Processing for ${projectExec.getClass} is currently supported.")
+      ProjectExecTransformer(projectExec.projectList, projectExec.child)
+    } else if (BackendsApiManager.getSettings.supportNativeInputFileNameExpr()) {
+      val addHint = AddTransformHintRule()
+      val newProjectList = projectExec.projectList.filterNot(isInputFileNameExpr)
+      val newProjectExec = ProjectExec(newProjectList, projectExec.child)
+      addHint.apply(newProjectExec)
+      if (TransformHints.isNotTransformable(newProjectExec)) {
+        // Project is still not transformable after remove `input_file_name` expressions.
+        projectExec
+      } else {
+        // the project with `input_file_name` expression should have at most
+        // one data source, reference:
+        // https://github.com/apache/spark/blob/e459674127e7b21e2767cc62d10ea6f1f941936c
+        // /sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L506
+        val scanChild = findScanNodes(projectExec)
+        if (scanChild.isEmpty || TransformHints.isNotTransformable(scanChild(0))) {
+          // It means
+          // 1. projectExec has `input_file_name` but no scan child.
+          // 2. It has scan child node but the scan node fallback.
+          projectExec
+        } else {
+          val replacedExprs = scala.collection.mutable.Map[String, AttributeReference]()
+          val newProjectList = projectExec.projectList.map {
+            expr => rewriteExpr(expr, replacedExprs).asInstanceOf[NamedExpression]
+          }
+          val newChild = addMetadataCol(projectExec.child, replacedExprs)
+          logDebug(s"Columnar Processing for ${projectExec.getClass} is currently supported.")
+          ProjectExecTransformer(newProjectList, newChild)
+        }
+      }
+    } else projectExec
+  }
+
+  override def offload(plan: SparkPlan): SparkPlan = plan match {
+    case p: ProjectExec =>
+      genProjectExec(p)
+    case other => other
+  }
```

Comment:
Is it possible to have different entrance in code between offloading regular projects and the ones with `input_file_name`?

e.g.,

```scala
case p: SparkPlan if TransformHints.isNotTransformable(p) => p
case p: ProjectExec if requestsInputFileName(p) => offloadProjectWithInputFileName(p)
case p: ProjectExec => ProjectExecTransformer(p.projectList, p.child)
```

Reply:
sure, i do a similiar refactor to put `offload project with inputFile related exprs` in seperate method